packages/keyring-controller/src/KeyringController.ts

```diff
@@ -1299,6 +1309,69 @@ export class KeyringController extends BaseController<
     return seedWords;
   }
 
+  /**
+   * Select a keyring and execute the given function with
+   * the selected keyring, as a mutually exclusive atomic
+   * operation.
+   *
+   * The method automatically persists changes at the end of the
+   * function execution, or rolls back the changes if an error
+   * is thrown.
+   *
+   * @param selector - Keyring selector object.
+   * @param fn - Function to execute with the selected keyring.
+   * @param options - Additional options.
+   * @param options.createIfMissing - Whether to create a new keyring if the selected one is missing.
+   * @param options.createWithData - Optional data to use when creating a new keyring.
+   * @returns Promise resolving to the result of the function execution.
+   */
+  async withKeyring<K extends EthKeyring<Json> = EthKeyring<Json>, T = void>(
+    selector: KeyringSelector,
+    fn: (keyring: K) => Promise<T>,
+    options:
+      | { createIfMissing?: false }
+      | { createIfMissing: true; createWithData?: unknown } = {
+      createIfMissing: false,
+    },
+  ): Promise<T> {
+    return this.#persistOrRollback(async () => {
+      let keyring: K | undefined;
+
+      if ('address' in selector) {
+        keyring = (await this.getKeyringForAccount(
+          normalize(selector.address) as Hex,
```

Comment:
Nit: Is this `normalize` call needed? It seems that the address is normalized immediately in `getKeyringForAccount`

Reply:
Hmm, I can't recall why I thought it was needed. I removed it in [d51ca05](https://github.com/MetaMask/core/pull/4197/commits/d51ca0504757b04c3d921d6f06f2d88475b85206)